tree/ntuple/v7/src/RPageStorageDaos.cxx

```diff
@@ -0,0 +1,462 @@
+
+#include <ROOT/RCluster.hxx>
+#include <ROOT/RClusterPool.hxx>
+#include <ROOT/RField.hxx>
+#include <ROOT/RLogger.hxx>
+#include <ROOT/RNTupleDescriptor.hxx>
+#include <ROOT/RNTupleModel.hxx>
+#include <ROOT/RNTupleZip.hxx>
+#include <ROOT/RPage.hxx>
+#include <ROOT/RPageAllocator.hxx>
+#include <ROOT/RPagePool.hxx>
+#include <ROOT/RDaos.hxx>
+#include <ROOT/RPageStorageDaos.hxx>
+
+#include <RVersion.h>
+#include <TError.h>
+
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <utility>
+#include <regex>
+
+namespace {
+/**
+  \brief Parse a DAOS pool URI of the form 'daos://pool-uuid/svc-replicas'.
+*/
+std::tuple<std::string /* poolUUID */, std::string /* svc */> ParseDaosPoolURI(std::string_view uri)
+{
+   std::regex re("daos://([[:xdigit:]-]+)/([[:digit:]:]+)");
+   std::cmatch m;
+   if (!std::regex_match(uri.data(), m, re))
+      throw std::runtime_error("Invalid DAOS pool URI.");
+   return { m[1], m[2] };
+}
+
+static constexpr std::uint64_t kDistributionKey = 0x5a3c69f0cafe4a11;
+static constexpr std::uint64_t kAttributeKey = 0x4243544b5344422d;
+
+static constexpr daos_obj_id_t kOidAnchor{std::uint64_t(-1), 0};
+static constexpr daos_obj_id_t kOidHeader{std::uint64_t(-2), 0};
+static constexpr daos_obj_id_t kOidFooter{std::uint64_t(-3), 0};
+} // namespace
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPageSinkDaos::RPageSinkDaos(std::string_view ntupleName, std::string_view path,
+   const RNTupleWriteOptions &options)
+   : RPageSink(ntupleName, options)
+   , fMetrics("RPageSinkRoot")
+   , fPageAllocator(std::make_unique<RPageAllocatorHeap>())
+{
+   auto args = ParseDaosPoolURI(path);
+   fPool = std::make_shared<RDaosPool>(std::get<0>(args), std::get<1>(args));
+   fContainer = std::unique_ptr<RDaosContainer>(new RDaosContainer(fPool, ntupleName, /*create =*/ true));
+}
+
+
+ROOT::Experimental::Detail::RPageSinkDaos::~RPageSinkDaos()
+{
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::CreateImpl(const RNTupleModel & /* model */)
+{
+   const auto &descriptor = fDescriptorBuilder.GetDescriptor();
+   auto szHeader = descriptor.SerializeHeader(nullptr);
+   auto buffer = std::unique_ptr<unsigned char[]>(new unsigned char[szHeader]);
+   descriptor.SerializeHeader(buffer.get());
+
+   auto zipBuffer = std::unique_ptr<unsigned char[]>(new unsigned char[szHeader]);
+   auto szZipHeader = fCompressor(buffer.get(), szHeader, fOptions.GetCompression(),
+      [&zipBuffer](const void *b, size_t n, size_t o){ memcpy(zipBuffer.get() + o, b, n); } );
+   WriteNTupleHeader(zipBuffer.get(), szZipHeader, szHeader);
+}
+
+
+ROOT::Experimental::RClusterDescriptor::RLocator
+ROOT::Experimental::Detail::RPageSinkDaos::CommitPageImpl(ColumnHandle_t columnHandle, const RPage &page)
+{
+   unsigned char *buffer = reinterpret_cast<unsigned char *>(page.GetBuffer());
```

Comment:
I think we should add general-purpose `SerializePage` and `DeserializePage` routines to factor out mem-to-disk and disk-to-mem page processing for the most common cases.  As Axel mentioned elsewhere, we currently have code snippet copies in several places across the page storage classes.  For a later PR though.

Reply:
Added TODO note.